src/modules/discovery/hosts.py

```diff
@@ -86,54 +113,32 @@ def __init__(self, event):
         self.event = event
 
     def execute(self):
-        # Scan any hosts that the user specified
+        # Scan any additional hosts that the user specified
         if config.remote or config.cidr:
             self.publish_event(HostScanEvent())
-        else:
-            # Discover cluster subnets, we'll scan all these hosts 
-            if self.is_azure_pod():
-                subnets, cloud = self.azure_metadata_discovery()
-            else:
-                subnets, cloud = self.traceroute_discovery()
-
-            for subnet in subnets:
-                logging.debug("From pod scanning subnet {0}/{1}".format(subnet[0], subnet[1]))
-                for ip in HostDiscoveryHelpers.generate_subnet(ip=subnet[0], sn=subnet[1]):
-                    self.publish_event(NewHostEvent(host=ip, cloud=cloud))
-
-            
-    def is_azure_pod(self):
-        try:
-            logging.debug("From pod attempting to access Azure Metadata API")
-            if requests.get("http://169.254.169.254/metadata/instance?api-version=2017-08-01", headers={"Metadata":"true"}, timeout=5).status_code == 200:
-                return True
-        except requests.exceptions.ConnectionError:
-            return False
-
-   # for pod scanning
-    def traceroute_discovery(self):
-        external_ip = requests.get("http://canhazip.com").text # getting external ip, to determine if cloud cluster
-        cloud = HostDiscoveryHelpers.get_cloud(external_ip)
-        logging.getLogger("scapy.runtime").setLevel(logging.ERROR) # disables scapy's warnings
-        from scapy.all import ICMP, IP, Ether, srp1
-
-        node_internal_ip = srp1(Ether() / IP(dst="google.com" , ttl=1) / ICMP(), verbose=0)[IP].src
-        return [ [node_internal_ip,"24"], ], external_ip
-
-    # quering azure's interface metadata api | works only from a pod
-    def azure_metadata_discovery(self):
-        logging.debug("From pod attempting to access azure's metadata")
-        machine_metadata = json.loads(requests.get("http://169.254.169.254/metadata/instance?api-version=2017-08-01", headers={"Metadata":"true"}).text)
-        address, subnet= "", ""
-        subnets = list()
-        for interface in machine_metadata["network"]["interface"]:
-            address, subnet = interface["ipv4"]["subnet"][0]["address"], interface["ipv4"]["subnet"][0]["prefix"]
-            logging.debug("From pod discovered subnet {0}/{1}".format(address, subnet if not config.quick else "24"))
-            subnets.append([address,subnet if not config.quick else "24"])
-
-            self.publish_event(AzureMetadataApi(cidr="{}/{}".format(address, subnet)))
-
-        return subnets, "Azure"
+
```

Comment:
I think this changes behaviour so that it will always scan the pod subnet even if the user specified remote / cidr. See https://github.com/aquasecurity/kube-hunter/pull/82 - we stopped it from doing that for the situation where you explicitly want to run tests against a remote address from inside a pod. (Imagine you have multiple Kubernetes clusters, you might want to simulate whether an attacker in a compromised pod can now get to your other clusters.)

Reply:
But what about when you know your cluster cidr, and want kubehunter to discover services on other nodes as well as the one he is running at? I think it's better to have a report for the whole cluster instead of you needing to specify explicitly each time the ip's. it's running on the node, it makes sense that these parameters will be considered as "extra" instead of a replacement

Reply:
Yes. In the #82 scenario the idea was to attack a specific IP address, and it was hard to view the results (especially with debug logging turned on) because there were so many tests based on being inside a pod going on. The default behaviour is to do all these local tests, so that is what most people’s experience will be. But if you go non-default by specifying a remote address / cidr I think it’s more likely you are explicitly trying to test the remote address rather than looking for every possible vulnerability we can find, locally or remotely. It’s also IMO more consistent with the normal, outside-the-pod behaviour - when we specify a remote address, that’s the only address we test against (we don’t also try to explore all the local interfaces).

Reply:
I have a suggestion, lmk what you think.
Although I don't want to massively change architecture, for consistent purposes, I believe the best solution would be to address the --pod differently.
We confuse this flag behavior a lot, Currently it does two things:
*  It publishes a RunningAsPodEvent 
*  by default it will run a local subnet scan, very similar to the --internal flag, if not the same.

We should really separate the two.
The normal behavior of this flag IMO should just be to publish the RunningAsPodEvent. 
If you want to scan internal node, also pass the --internal flag. So for example:
`./kube-hunter.py --pod --internal`
This will make it consistent with normal internal scanning, 
`./kube-hunter.py --internal`

And if you want to run a different scanning method. just use:
`./kube-hunter.py --pod --remote `

of course the flags can be aggregated, as passing:
`./kube-hunter.py --pod --internal --remote 0.0.0.0 --cidr 1.1.1.1/24`
will use all scanning methods.

Essentially, this default behavior of scanning the whole node is confusing. using the --internal flag should be the only method to initiate an internal scan, either as a pod or in normal scan
@lizrice

Reply:
Sorry for slow response on this, I missed this comment somehow. Yes. this suggestion makes sense to me, it would make more sense for --pod to just mean one thing. (Would be even better if we could somehow auto-detect that we're running inside a pod, but we can handle that separately).

Reply:
After a discussion with @mcherny  I returned the behavior to be like before, so if you specify --cidr or --remote, the pod internal scanning is not taking place. in the future if we want to change this, we should probably do this in another PR, this PR made enough big changes.. :)